### ISSUE TITLE
time offset between tsync/tconnect record and NS time parse

### ISSUE THREAD

**Issue by jyaw** [jwoglom/tconnectsync#47]
Observed a couple problems with my initial tconnectsync attempts. I'll paste the full content of relevant errors in the next comment. Highlighting 2 items here...

1. Calculated "5 hours" offset between the similar log time and CIQ time? Not sure why this is hapenning. My timezone on my linux system and the date command are set correctly (central time)... (e.g. 2022-08-13 20:38:42 INFO     Last bolus from t:connect CIQ: 2022-08-13T19:46:15 (5 hours, 52 minutes ago) )
2. NS latest entry update shows a ISO-8601 parsing error, note that the NS site is a brand new site with not records. Tried adding a fake bolus manually and that didn't seem to change the error.

raise ApiException(latest.status_code, "Nightscout last_uploaded_entry response: %s" % latest.text)
tconnectsync.api.common.ApiException: Nightscout last_uploaded_entry response: 
Error: Cannot parse 2022-08-12+20:38:37 as a valid ISO-8601 date

**Comment by jyaw** [comment]
Full error text: 
2022-08-13 20:38:37 INFO     Enabled features: BASAL, BOLUS
2022-08-13 20:38:37 INFO     Logged in to AndroidApi successfully (expiration: 2022-08-14T09:38:37.657Z, in 7 hours, 59 minutes)
2022-08-13 20:38:37 INFO     New reported t:connect data. (event index: 825914 last: None)
2022-08-13 20:38:37 INFO     Downloading t:connect ControlIQ data
2022-08-13 20:38:37 INFO     Logging in to ControlIQApi...
2022-08-13 20:38:38 INFO     Reported tconnect software version: t:connect 7.14.0.1
2022-08-13 20:38:40 INFO     Logged in to ControlIQApi successfully (expiration: 2022-08-14T09:38:38.752Z, in 7 hours, 59 minutes)
2022-08-13 20:38:41 INFO     Downloading t:connect therapy_events
2022-08-13 20:38:42 WARNING  No last CGM reading is able to be determined from CIQ
2022-08-13 20:38:42 INFO     Last bolus from t:connect CIQ: 2022-08-13T19:46:15 (5 hours, 52 minutes ago)
2022-08-13 20:38:43 INFO     Last basal event from CIQ: {'time': '2022-08-13 19:52:44-05:00', 'delivery_type': 'algorithmDelivery', 'duration_mins': 3.2, 'basal_rate': 1.383}
Traceback (most recent call last):
  File "/home/pi/.local/bin/tconnectsync", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/tconnectsync/__init__.py", line 87, in main
    sys.exit(u.process(tconnect, nightscout, time_start, time_end, args.pretend, features=args.features))
  File "/home/pi/.local/lib/python3.9/site-packages/tconnectsync/autoupdate.py", line 48, in process
    added = process_time_range(tconnect, nightscout, time_start, time_end, pretend, features=features)
  File "/home/pi/.local/lib/python3.9/site-packages/tconnectsync/process.py", line 143, in process_time_range
    added += ns_write_basal_events(nightscout, basalEvents, pretend=pretend, time_start=time_start, time_end=time_end)
  File "/home/pi/.local/lib/python3.9/site-packages/tconnectsync/sync/basal.py", line 109, in ns_write_basal_events
    last_upload = nightscout.last_uploaded_entry(BASAL_EVENTTYPE, time_start=time_start, time_end=time_end)
  File "/home/pi/.local/lib/python3.9/site-packages/tconnectsync/nightscout.py", line 64, in last_uploaded_entry
    raise ApiException(latest.status_code, "Nightscout last_uploaded_entry response: %s" % latest.text)
tconnectsync.api.common.ApiException: Nightscout last_uploaded_entry response: <html>
  <head>
    <meta charset='utf-8'>
    <title>Error: Cannot parse 2022-08-12+20:38:37 as a valid ISO-8601 date</title>
    <style>* {
  margin: 0;
  padding: 0;
  outline: 0;
}

body {
  padding: 80px 100px;
  font: 13px "Helvetica Neue", "Lucida Grande", "Arial";
  background: #ECE9E9 -webkit-gradient(linear, 0% 0%, 0% 100%, from(#fff), to(#ECE9E9));
  background: #ECE9E9 -moz-linear-gradient(top, #fff, #ECE9E9);
  background-repeat: no-repeat;
  color: #555;
  -webkit-font-smoothing: antialiased;
}
h1, h2 {
  font-size: 22px;
  color: #343434;
}
h1 em, h2 em {
  padding: 0 5px;
  font-weight: normal;
}
h1 {
  font-size: 60px;
}
h2 {
  margin-top: 10px;
}
ul li {
  list-style: none;
}
#stacktrace {
  margin-left: 60px;
}
</style>
  </head>
  <body>
    <div id="wrapper">
      <h1>Connect</h1>
      <h2><em>500</em> Error: Cannot parse 2022-08-12+20:38:37 as a valid ISO-8601 date</h2>
      <ul id="stacktrace"><li> &nbsp; &nbsp;at /app/lib/server/query.js:71:17</li><li> &nbsp; &nbsp;at Array.forEach (&lt;anonymous&gt;)</li><li> &nbsp; &nbsp;at enforceDateFilter (/app/lib/server/query.js:62:28)</li><li> &nbsp; &nbsp;at create (/app/lib/server/query.js:120:5)</li><li> &nbsp; &nbsp;at query_for (/app/lib/server/treatments.js:121:12)</li><li> &nbsp; &nbsp;at Function.list (/app/lib/server/treatments.js:115:13)</li><li> &nbsp; &nbsp;at /app/lib/api/treatments/index.js:94:22</li><li> &nbsp; &nbsp;at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)</li><li> &nbsp; &nbsp;at next (/app/node_modules/express/lib/router/route.js:137:13)</li><li> &nbsp; &nbsp;at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)</li></ul>
    </div>
  </body>
</html>
 (HTTP 500)

**Comment by t1diotac** [comment]
It appears I'm getting a similar error regarding parsing the date as a valid ISO-8601 date.
```
2022-08-18 20:38:23 INFO     Enabled features: BASAL, BOLUS, PUMP_EVENTS
Processing data between 2022-08-17 20:38:23.528173 and 2022-08-18 20:38:23.528173 
2022-08-18 20:38:23 INFO     Downloading t:connect ControlIQ data
2022-08-18 20:38:23 INFO     Logging in to ControlIQApi...
2022-08-18 20:38:23 INFO     Reported tconnect software version: t:connect 7.14.0.1
2022-08-18 20:38:27 INFO     Logged in to ControlIQApi successfully (expiration: 2022-08-19T10:38:24.005Z, in 7 hours, 59 minutes)
2022-08-18 20:38:28 INFO     Downloading t:connect therapy_events
2022-08-18 20:38:29 WARNING  No last CGM reading is able to be determined from CIQ
2022-08-18 20:38:29 INFO     Last bolus from t:connect CIQ: 2022-08-18T19:41:42 (6 hours, 56 minutes ago)
2022-08-18 20:38:29 INFO     Last basal event from CIQ: {'time': '2022-08-18 19:55:52-04:00', 'delivery_type': 'algorithmDelivery', 'duration_mins': 30.066666666666666, 'basal_rate': 0.25}
Traceback (most recent call last):
  File "/home/health/.local/bin/tconnectsync", line 8, in <module>
    sys.exit(main())
  File "/home/health/.local/lib/python3.9/site-packages/tconnectsync/__init__.py", line 90, in main
    added = process_time_range(tconnect, nightscout, time_start, time_end, args.pretend, features=args.features)
  File "/home/health/.local/lib/python3.9/site-packages/tconnectsync/process.py", line 143, in process_time_range
    added += ns_write_basal_events(nightscout, basalEvents, pretend=pretend, time_start=time_start, time_end=time_end)
  File "/home/health/.local/lib/python3.9/site-packages/tconnectsync/sync/basal.py", line 109, in ns_write_basal_events
    last_upload = nightscout.last_uploaded_entry(BASAL_EVENTTYPE, time_start=time_start, time_end=time_end)
  File "/home/health/.local/lib/python3.9/site-packages/tconnectsync/nightscout.py", line 64, in last_uploaded_entry
    raise ApiException(latest.status_code, "Nightscout last_uploaded_entry response: %s" % latest.text)
tconnectsync.api.common.ApiException: Nightscout last_uploaded_entry response: <html>
  <head>
    <meta charset='utf-8'> 
    <title>Error: Cannot parse 2022-08-17+20:38:23 as a valid ISO-8601 date</title>
    <style>* {
  margin: 0;
  padding: 0;
  outline: 0;
}

body {
  padding: 80px 100px;
  font: 13px "Helvetica Neue", "Lucida Grande", "Arial";
  background: #ECE9E9 -webkit-gradient(linear, 0% 0%, 0% 100%, from(#fff), to(#ECE9E9));
  background: #ECE9E9 -moz-linear-gradient(top, #fff, #ECE9E9);
  background-repeat: no-repeat;
  color: #555;
  -webkit-font-smoothing: antialiased;
}
h1, h2 {
  font-size: 22px;
  color: #343434;
}
h1 em, h2 em {
  padding: 0 5px;
  font-weight: normal;
}
h1 {
  font-size: 60px;
}
h2 {
  margin-top: 10px;
}
ul li {
  list-style: none;
}
#stacktrace {
  margin-left: 60px;
}
</style>
  </head>
  <body>
    <div id="wrapper">
      <h1>Connect</h1>
      <h2><em>500</em> Error: Cannot parse 2022-08-17+20:38:23 as a valid ISO-8601 date</h2>
      <ul id="stacktrace"><li> &nbsp; &nbsp;at /app/lib/server/query.js:71:17</li><li> &nbsp; &nbsp;at Array.forEach (&lt;anonymous&gt;)</li><li> &nbsp; &nbsp;at enforceDateFilter (/app/lib/server/query.js:62:28)</li><li> &nbsp; &nbsp;at create (/app/lib/server/query.js:120:5)</li><li> &nbsp; &nbsp;at query_for (/app/lib/server/treatments.js:121:12)</li><li> &nbsp; &nbsp;at Function.list (/app/lib/server/treatments.js:115:13)</li><li> &nbsp; &nbsp;at /app/lib/api/treatments/index.js:94:22</li><li> &nbsp; &nbsp;at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)</li><li> &nbsp; &nbsp;at next (/app/node_modules/express/lib/router/route.js:137:13)</li><li> &nbsp; &nbsp;at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)</li></ul>
    </div>
  </body>
</html>
 (HTTP 500)
```

**Comment by rstutsman** [comment]
I'm running into the same issue on Ubuntu 20.04 but it is working macos. The issue seems to stem from `urljoin` of the `dateFilter` returned by `time_range` in `last_uploaded_activity`. This is just a guess, but it looks like the space in put into the date format in `time_range` is getting encoded as a `+`. It seems like ISO-8601 doesn't specify the space anyway, so maybe it makes sense to remove it?

**Comment by jyaw** [comment]
Thanks @rstutsman! I figured it had to do with the date filter... will give this a check on my instance sometime today.

**Comment by jyaw** [comment]
Hey I'm getting a different unknown therapy event error now before I hit the previous iso format time error, can't really verify yet. Will update if I'm able to check it out, hopefully soon.

**Comment by ramboozer** [comment]
`pi@magicmirroros:~ $ tconnectsync --auto-update
2022-08-20 20:17:19 INFO     Enabled features: BASAL, BOLUS
Starting auto-update between 2022-08-19 20:17:19.923273 and 2022-08-20 20:17:19.923273
2022-08-20 20:17:20 INFO     Logged in to AndroidApi successfully (expiration: 2022-08-21T08:17:20.401Z, in 7 hours, 59 minutes)
2022-08-20 20:17:20 INFO     New reported t:connect data. (event index: 817956 last: None)
2022-08-20 20:17:20 INFO     Downloading t:connect ControlIQ data
2022-08-20 20:17:20 INFO     Logging in to ControlIQApi...
2022-08-20 20:17:21 INFO     Reported tconnect software version: t:connect 7.14.0.1
2022-08-20 20:17:24 INFO     Logged in to ControlIQApi successfully (expiration: 2022-08-21T08:17:21.513Z, in 7 hours, 59 minutes)
2022-08-20 20:17:25 INFO     Downloading t:connect therapy_events
Traceback (most recent call last):
  File "/usr/local/bin/tconnectsync", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/dist-packages/tconnectsync/__init__.py", line 87, in main
    sys.exit(u.process(tconnect, nightscout, time_start, time_end, args.pretend, features=args.features))
  File "/usr/local/lib/python3.7/dist-packages/tconnectsync/autoupdate.py", line 48, in process
    added = process_time_range(tconnect, nightscout, time_start, time_end, pretend, features=features)
  File "/usr/local/lib/python3.7/dist-packages/tconnectsync/process.py", line 67, in process_time_range
    ciqBolusData, ciqReadingData = split_therapy_events(ciqTherapyEventsData)
  File "/usr/local/lib/python3.7/dist-packages/tconnectsync/parser/ciq_therapy_events.py", line 12, in split_therapy_events
    event = TConnectEntry.parse_therapy_event(e)
  File "/usr/local/lib/python3.7/dist-packages/tconnectsync/parser/tconnect.py", line 204, in parse_therapy_event
    raise UnknownTherapyEventException(data)
tconnectsync.parser.tconnect.UnknownTherapyEventException: Unknown therapy event type:
`

**Comment by jyaw** [comment]
@Ya that’s what I’m seeing as well!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jyaw** [comment]
It looks like when process.py calls parse_therapy_event in parser/tconnect.py, the type for the first few of my events is 'BG', rather than 'Bolus' or 'CGM', so it throws a UnknownTherapyEventException... not related to this time issue, so I'll open a new issue if it doesnt exist yet...

**Comment by jyaw** [comment]
Verified it works, will leave the issue open since the PR hasn't been approved...

**Comment by rstutsman** [comment]
@jyaw I ran into the UnknownTherapyEventException for the same reason as you (`type == BG`), but before I could track down what was the issue disappeared. If someone catches that issue in action, it would be great to capture it in an issue. I tweaked the output for UnknownTherapyEventException in a posted PR since it was missing a format string element that prevented it from outputting the full details of the event that triggered it.
